state/charm.go

```diff
@@ -1121,11 +1121,17 @@ func (st *State) AddCharmMetadata(info CharmInfo) (*Charm, error) {
 	}
 
 	buildTxn := func(attempt int) ([]txn.Op, error) {
+		ops, err := insertCharmOps(st, info)
+		if err == nil {
+			return ops, nil
+		}
+		if err != nil && !errors.Is(err, errors.AlreadyExists) {
```

Comment:
Thanks for this. I know I've fixed a few of these cases over the years. It's like playing whack-a-mole. 

There is a ci test to cover the scenario: https://github.com/juju/juju/blob/8951305ad67a5d467f515b94384247caf33cdc8c/tests/suites/deploy/deploy_bundles.sh#L313  

The test is passing on 3.5 and 3.6. With new info, are there any updates which could be made to this test?

Reply:
quoting @wallyworld: "the bundle is picked apart client side and individual deploy commands are made one by one as if the user had typed them. so in effect it happens in parallel server side".. so i do not think a bundle deploy would've caught this race condition. A terraform provider test would, though.